In [7]:
#Wilson Song
#Assigntment 04

from math import sin, cos, cosh, sinh, exp, factorial



x1 = 1/4
x2 = 1/2
x3 = 1
x4 = 2

n1 = 4
n2 = 16
n3 = 64



#expression to calculate derivative of sin

def sin_derivative(k): #module 4 of k to find the iteration that sin is in
    if k % 4 == 0:  
        return sin(0) # sin(0)=0
    elif k % 4 == 1:
        return cos(0) # cos(0)=1
    elif k % 4 == 2:
        return (-1) * sin(0) #-sin(0)=0
    elif k % 4 == 3:
        return (-1) * cos(0) #-cos(0)=-1


# expression to calculate derivative of exp

def exp_derivative(k):
    return exp(0)


# experession to calculate derivative of cosh


def cosh_derivative(k):
    if k % 2 == 0:  
        return 1 * cosh(0)
    elif k % 2 == 1:
        return 1 * sinh(0)
    


# taylor polynomial for exp

def taylor_sin(x, n):
    result = 0
    for k in range(n+1):
        result = result + (sin_derivative(k) / factorial(k)) * pow(x, k)
    return result


# taylor polynomial for exp


def taylor_exp(x, n):
    result = 0
    for k in range(n+1):
        result = result + (exp_derivative(k) / factorial(k)) * pow(x, k)
    return result


# taylor polynomial for cosh


def taylor_cosh(x, n):
    result = 0
    for k in range(n+1):
        result = result + ((cosh_derivative(k) / factorial(k)) * pow(x, k))
    return result

#sin(x)

sinx1_exact = sin(x1)
print("Exact Value of sin(1/4): "+str(sinx1_exact))

sinx1_n1_poly=taylor_sin(x1,n1)
print("Sin(1/4) with n=4:"+str(sinx1_n1_poly))
print("Relative Error:"+str((sinx1_n1_poly-sinx1_exact)/sinx1_exact))
print("Abs Error:"+str(abs((sinx1_n1_poly-sinx1_exact)/sinx1_exact)))
sinx1_n2_poly=taylor_sin(x1,n2)
print("Sin(1/4) with n=16:"+str(sinx1_n2_poly))
print("Relative Error:"+str((sinx1_n2_poly-sinx1_exact)/sinx1_exact))
print("Abs Error:"+str(abs((sinx1_n2_poly-sinx1_exact)/sinx1_exact)))
sinx1_n3_poly=taylor_sin(x1,n3)
print("Sin(1/4) with n=64:"+str(sinx1_n3_poly))
print("Relative Error:"+str((sinx1_n3_poly-sinx1_exact)/sinx1_exact))
print("Abs Error:"+str(abs((sinx1_n3_poly-sinx1_exact)/sinx1_exact)))

print("\n")
sinx2_exact = sin(x2)
print("Exact Value of sin(1/2): "+str(sinx2_exact))
sinx2_n1_poly=taylor_sin(x2,n1)
print("Sin(1/2) with n=4:"+str(sinx2_n1_poly))
print("Relative Error:"+str((sinx2_n1_poly-sinx2_exact)/sinx2_exact))
print("Abs Error:"+str(abs((sinx2_n1_poly-sinx2_exact)/sinx2_exact)))
sinx2_n2_poly=taylor_sin(x2,n2)
print("Sin(1/2) with n=16:"+str(sinx2_n2_poly))
print("Relative Error:"+str((sinx2_n2_poly-sinx2_exact)/sinx2_exact))
print("Abs Error:"+str(abs((sinx2_n2_poly-sinx2_exact)/sinx2_exact)))
sinx2_n3_poly=taylor_sin(x2,n3)
print("Sin(1/2) with n=64:"+str(sinx2_n3_poly))
print("Abs Error:"+str((sinx2_n3_poly-sinx2_exact)/sinx2_exact))
print("Abs Error:"+str(abs((sinx2_n3_poly-sinx2_exact)/sinx2_exact)))

print("\n")
sinx3_exact = sin(x3)
print("Exact Value of sin(1): "+str(sinx3_exact))
sinx3_n1_poly=taylor_sin(x3,n1)
print("Sin(1) with n=4:"+str(sinx3_n1_poly))
print("Relative Error:"+str((sinx3_n1_poly-sinx3_exact)/sinx3_exact))
print("Abs Error:"+str(abs((sinx3_n1_poly-sinx3_exact)/sinx3_exact)))
sinx3_n2_poly=taylor_sin(x3,n2)
print("Sin(1) with n=16:"+str(sinx3_n2_poly))
print("Relative Error:"+str((sinx3_n2_poly-sinx3_exact)/sinx3_exact))
print("Abs Error:"+str(abs((sinx3_n2_poly-sinx3_exact)/sinx3_exact)))
sinx3_n3_poly=taylor_sin(x3,n3)
print("Sin(1) with n=64:"+str(sinx3_n3_poly))
print("Relative Error:"+str((sinx3_n3_poly-sinx3_exact)/sinx3_exact))
print("Abs Error:"+str(abs((sinx3_n3_poly-sinx3_exact)/sinx3_exact)))

print("\n")
sinx4_exact = sin(x4)
print("Exact Value of sin(2): "+str(sinx4_exact))
sinx4_n1_poly=taylor_sin(x4,n1)
print("Sin(2) with n=4:"+str(sinx4_n1_poly))
print("Relative Error:"+str((sinx4_n1_poly-sinx4_exact)/sinx4_exact))
print("Abs Error:"+str(abs((sinx4_n1_poly-sinx4_exact)/sinx4_exact)))
sinx4_n2_poly=taylor_sin(x4,n2)
print("Sin(2) with n=16:"+str(sinx4_n2_poly))
print("Relative Error:"+str((sinx4_n2_poly-sinx4_exact)/sinx4_exact))
print("Abs Error:"+str(abs((sinx4_n2_poly-sinx4_exact)/sinx4_exact)))
sinx4_n3_poly=taylor_sin(x4,n3)
print("Sin(2) with n=64:"+str(sinx4_n3_poly))
print("Relative Error:"+str((sinx4_n3_poly-sinx4_exact)/sinx4_exact))
print("Abs Error:"+str(abs((sinx4_n3_poly-sinx4_exact)/sinx4_exact)))

#exp(x)
print("\n")
expx1_exact = exp(x1)
print("Exact Value of exp(1/4): "+str(expx1_exact))
expx1_n1_poly=taylor_exp(x1,n1)
print("Exp(1/4) with n=4:"+str(expx1_n1_poly))
print("Relative Error:"+str((expx1_n1_poly-expx1_exact)/expx1_exact))
print("Abs Error:"+str(abs((expx1_n1_poly-expx1_exact)/expx1_exact)))
expx1_n2_poly=taylor_exp(x1,n2)
print("Exp(1/4) with n=16:"+str(expx1_n2_poly))
print("Relative Error:"+str((expx1_n2_poly-expx1_exact)/expx1_exact))
print("Abs Error:"+str(abs((expx1_n2_poly-expx1_exact)/expx1_exact)))
expx1_n3_poly=taylor_exp(x1,n3)
print("Exp(1/4) with n=64:"+str(expx1_n3_poly))
print("Relative Error:"+str((expx1_n3_poly-expx1_exact)/expx1_exact))
print("Abs Error:"+str(abs((expx1_n3_poly-expx1_exact)/expx1_exact)))

print("\n")
expx2_exact = exp(x2)
print("Exact Value of exp(1/2): "+str(expx2_exact))
expx2_n1_poly=taylor_exp(x2,n1)
print("Exp(1/2) with n=4:"+str(expx2_n1_poly))
print("Relative Error:"+str((expx2_n1_poly-expx2_exact)/expx2_exact))
print("Abs Error:"+str(abs((expx2_n1_poly-expx2_exact)/expx2_exact)))
expx2_n2_poly=taylor_exp(x2,n2)
print("Exp(1/2) with n=16:"+str(expx2_n2_poly))
print("Relative Error:"+str((expx2_n2_poly-expx2_exact)/expx2_exact))
print("Abs Error:"+str(abs((expx2_n2_poly-expx2_exact)/expx2_exact)))
expx2_n3_poly=taylor_exp(x2,n3)
print("Exp(1/2) with n=64:"+str(expx2_n3_poly))
print("Relative Error:"+str((expx2_n3_poly-expx2_exact)/expx2_exact))
print("Abs Error:"+str(abs((expx2_n3_poly-expx2_exact)/expx2_exact)))

print("\n")
expx3_exact = exp(x3)
print("Exact Value of exp(1): "+str(expx3_exact))
expx3_n1_poly=taylor_exp(x3,n1)
print("Exp(1) with n=4:"+str(expx3_n1_poly))
print("Relative Error:"+str((expx3_n1_poly-expx3_exact)/expx3_exact))
print("Abs Error:"+str(abs((expx3_n1_poly-expx3_exact)/expx3_exact)))
expx3_n2_poly=taylor_exp(x3,n2)
print("Exp(1) with n=16:"+str(expx3_n2_poly))
print("Relative Error:"+str((expx3_n2_poly-expx3_exact)/expx3_exact))
print("Abs Error:"+str(abs((expx3_n2_poly-expx3_exact)/expx3_exact)))
expx3_n3_poly=taylor_exp(x3,n3)
print("Exp(1) with n=64:"+str(expx3_n3_poly))
print("Relative Error:"+str((expx3_n3_poly-expx3_exact)/expx3_exact))
print("Abs Error:"+str(abs((expx3_n3_poly-expx3_exact)/expx3_exact)))

print("\n")
expx4_exact = exp(x4)
print("Exact Value of exp(2): "+str(expx4_exact))
expx4_n1_poly=taylor_exp(x4,n1)
print("Exp(2) with n=4:"+str(expx4_n1_poly))
print("Relative Error:"+str((expx4_n1_poly-expx4_exact)/expx4_exact))
print("Abs Error:"+str(abs((expx4_n1_poly-expx4_exact)/expx4_exact)))
expx4_n2_poly=taylor_exp(x4,n2)
print("Exp(2) with n=16:"+str(expx4_n2_poly))
print("Relative Error:"+str((expx4_n2_poly-expx4_exact)/expx4_exact))
print("Abs Error:"+str(abs((expx4_n2_poly-expx4_exact)/expx4_exact)))
expx4_n3_poly=taylor_exp(x4,n3)
print("Exp(2) with n=64:"+str(expx4_n3_poly))
print("Relative Error:"+str((expx4_n3_poly-expx4_exact)/expx4_exact))
print("Abs Error:"+str(abs((expx4_n3_poly-expx4_exact)/expx4_exact)))
#cosh(x)

print("\n")
coshx1_exact = cosh(x1)
print("Exact Value of cosh(1/4): "+str(coshx1_exact))
coshx1_n1_poly=taylor_cosh(x1,n1)
print("Cosh(1/4) with n=4:"+str(coshx1_n1_poly))
print("Relative Error:"+str((coshx1_n1_poly-coshx1_exact)/coshx1_exact))
print("Abs Error:"+str(abs((coshx1_n1_poly-coshx1_exact)/coshx1_exact)))
coshx1_n2_poly=taylor_cosh(x1,n2)
print("Cosh(1/4) with n=16:"+str(coshx1_n2_poly))
print("Relative Error:"+str((coshx1_n2_poly-coshx1_exact)/coshx1_exact))
print("Abs Error:"+str(abs((coshx1_n2_poly-coshx1_exact)/coshx1_exact)))
coshx1_n3_poly=taylor_cosh(x1,n3)
print("Cosh(1/4) with n=64:"+str(coshx1_n3_poly))
print("Relative Error:"+str((coshx1_n3_poly-coshx1_exact)/coshx1_exact))
print("Abs Error:"+str(abs((coshx1_n3_poly-coshx1_exact)/coshx1_exact)))

print("\n")
coshx2_exact = cosh(x2)
print("Exact Value of cosh(1/2): "+str(coshx2_exact))
coshx2_n1_poly=taylor_cosh(x2,n1)
print("Cosh(1/2) with n=4:"+str(coshx2_n1_poly))
print("Relative Error:"+str((coshx2_n1_poly-coshx2_exact)/coshx2_exact))
print("Abs Error:"+str(abs((coshx2_n1_poly-coshx2_exact)/coshx2_exact)))
coshx2_n2_poly=taylor_cosh(x2,n2)
print("Cosh(1/2) with n=16:"+str(coshx2_n2_poly))
print("Relative Error:"+str((coshx2_n2_poly-coshx2_exact)/coshx2_exact))
print("Abs Error:"+str(abs((coshx2_n2_poly-coshx2_exact)/coshx2_exact)))
coshx2_n3_poly=taylor_cosh(x2,n3)
print("Cosh(1/2) with n=64:"+str(coshx2_n3_poly))
print("Relative Error:"+str((coshx2_n3_poly-coshx2_exact)/coshx2_exact))
print("Abs Error:"+str(abs((coshx2_n3_poly-coshx2_exact)/coshx2_exact)))

print("\n")
coshx3_exact = cosh(x3)
print("Exact Value of cosh(1): "+str(coshx3_exact))
coshx3_n1_poly=taylor_cosh(x3,n1)
print("Cosh(1) with n=4:"+str(coshx3_n1_poly))
print("Relative Error:"+str((coshx3_n1_poly-coshx3_exact)/coshx3_exact))
print("Abs Error:"+str(abs((coshx3_n1_poly-coshx3_exact)/coshx3_exact)))
coshx3_n2_poly=taylor_cosh(x3,n2)
print("Cosh(1) with n=16:"+str(coshx3_n2_poly))
print("Relative Error:"+str((coshx3_n2_poly-coshx3_exact)/coshx3_exact))
print("Abs Error:"+str(abs((coshx3_n2_poly-coshx3_exact)/coshx3_exact)))
coshx3_n3_poly=taylor_cosh(x3,n3)
print("Cosh(1) with n=64:"+str(coshx3_n3_poly))
print("Relative Error:"+str((coshx3_n3_poly-coshx3_exact)/coshx3_exact))
print("Abs Error:"+str(abs((coshx3_n3_poly-coshx3_exact)/coshx3_exact)))

print("\n")
coshx4_exact = cosh(x4)
print("Exact Value of cosh(2): "+str(coshx4_exact))
coshx4_n1_poly=taylor_cosh(x4,n1)
print("Cosh(2) with n=4:"+str(coshx4_n1_poly))
print("Relative Error:"+str((coshx4_n1_poly-coshx4_exact)/coshx4_exact))
print("Abs Error:"+str(abs((coshx4_n1_poly-coshx4_exact)/coshx4_exact)))
coshx4_n2_poly=taylor_cosh(x4,n2)
print("Cosh(2) with n=16:"+str(coshx4_n2_poly))
print("Relative Error:"+str((coshx4_n2_poly-coshx4_exact)/coshx4_exact))
print("Abs Error:"+str(abs((coshx4_n2_poly-coshx4_exact)/coshx4_exact)))
coshx4_n3_poly=taylor_cosh(x4,n3)
print("Cosh(2) with n=64:"+str(coshx4_n3_poly))
print("Relative Error:"+str((coshx4_n3_poly-coshx4_exact)/coshx4_exact))
print("Abs Error:"+str(abs((coshx4_n3_poly-coshx4_exact)/coshx4_exact)))







Exact Value of sin(1/4): 0.24740395925452294
Sin(1/4) with n=4:0.24739583333333334
Relative Error:-3.2844749995430756e-05
Abs Error:3.2844749995430756e-05
Sin(1/4) with n=16:0.24740395925452294
Relative Error:0.0
Abs Error:0.0
Sin(1/4) with n=64:0.24740395925452294
Relative Error:0.0
Abs Error:0.0


Exact Value of sin(1/2): 0.479425538604203
Sin(1/2) with n=4:0.4791666666666667
Relative Error:-0.000539962761036883
Abs Error:0.000539962761036883
Sin(1/2) with n=16:0.479425538604203
Relative Error:0.0
Abs Error:0.0
Sin(1/2) with n=64:0.479425538604203
Abs Error:0.0
Abs Error:0.0


Exact Value of sin(1): 0.8414709848078965
Sin(1) with n=4:0.8333333333333334
Relative Error:-0.00967074518489894
Abs Error:0.00967074518489894
Sin(1) with n=16:0.8414709848078937
Relative Error:-3.2984590219667964e-15
Abs Error:3.2984590219667964e-15
Sin(1) with n=64:0.8414709848078965
Relative Error:0.0
Abs Error:0.0


Exact Value of sin(2): 0.9092974268256817
Sin(2) with n=4:0.6666666666666667
Relative Error:

|Function|X value|N value|Taylor polynomial Value|Relative error|Absolute error|Exact Value|
|---|---|---|---|---|---|---|
|Sin|1/4|4|4:0.24739583333333334|-3.2844749995430756e-05|3.2844749995430756e-05|0.24740395925452294|
|Sin|1/4|16|0.24740395925452294|0.0|0.0|0.24740395925452294|
|Sin|1/4|64|0.24740395925452294|0.0|0.0|0.24740395925452294|
|---|---|---|---|---|---|---|
|Sin|1/2|4|0.4791666666666667|0.93677849016843|0.93677849016843|0.479425538604203|
|Sin|1/2|16|0.479425538604203|0.0|0.0|0.479425538604203|
|Sin|1/2|64|0.479425538604203|0.0|0.0|0.479425538604203|
|---|---|---|---|---|---|---|
|Sin|1|4|0.8333333333333334|-0.00967074518489894|0.00967074518489894|0.8414709848078965|
|Sin|1|16|0.8414709848078937|-3.2984590219667964e-15|3.2984590219667964e-15|0.8414709848078965|
|Sin|1|64|0.8414709848078965|0.0|0.0|0.8414709848078965|
|---|---|---|---|---|---|---|
|Sin|2|4|0.6666666666666667|-0.26683321980358893|0.26683321980358893|0.9092974268256817|
|Sin|2|16|0.9092974264614476|-4.005664995743715e-10|4.005664995743715e-10|0.9092974268256817|
|Sin|2|64|0.9092974268256817|0.0|0.0|0.9092974268256817|
|---|---|---|---|---|---|---|
|Exp|1/4|4|1.2840169270833335|-6.6117105608491115e-06|6.6117105608491115e-06|1.2840254166877414|
|Exp|1/4|16|1.2840254166877418|3.458570243847902e-16|3.458570243847902e-16|1.2840254166877414|
|Exp|1/4|64|1.2840254166877418|3.458570243847902e-16|3.458570243847902e-16|1.2840254166877414|
|---|---|---|---|---|---|---|
|Exp|1/2|4|1.6484375|-0.00017211562995586948|0.00017211562995586948|1.6487212707001282|
|Exp|1/2|16|1.6487212707001278|-2.693537214216206e-16|2.693537214216206e-16|1.6487212707001282|
|Exp|1/2|64|1.6487212707001278|-2.693537214216206e-16|2.693537214216206e-161.6487212707001282|
|---|---|---|---|---|---|---|
|Exp|1|4|2.708333333333333|-0.003659846827343768|0.003659846827343768|2.718281828459045|
|Exp|1|16|2.718281828459043|-8.168564517495421e-16|8.168564517495421e-16|2.718281828459045|
|Exp|1|65|2.7182818284590455|1.6337129034990842e-16|1.6337129034990842e-16|2.718281828459045|
|---|---|---|---|---|---|---|
|Exp|2|4|7.0|-3.291240982792398e-07|3.291240982792398e-07|7.38905609893065|
|Exp|2|16|7.389056098516415|-5.606071347422176e-11|5.606071347422176e-11|7.38905609893065|
|Exp|2|64|7.389056098930649|-2.4040375598952704e-16|2.4040375598952704e-16|7.38905609893065|
|---|---|---|---|---|---|---|
|Cosh|1/4|4|1.0314127604166667|-3.291240982792398e-07|3.291240982792398e-07|1.0314130998795732|
|Cosh|1/4|16|1.0314130998795734|2.1528193209002e-16|2.1528193209002e-16|1.0314130998795732|
|Cosh|1/4|64|1.0314130998795734|2.1528193209002e-16|2.1528193209002e-16|1.0314130998795732|
|---|---|---|---|---|---|---|
|Cosh|1/2|4|1.1276041666666667|-1.9331356661309134e-05|1.9331356661309134e-05|1.1276259652063807|
|Cosh|1/2|16|1.1276259652063807|0.0|0.0|1.1276259652063807|
|Cosh|1/2|64|1.1276259652063807|0.0|0.0|1.1276259652063807|
|---|---|---|---|---|---|---|
|Cosh|1|4|1.5416666666666667|-0.0009163281015099182|0.0009163281015099182|1.5430806348152437|
|Cosh|1|16|1.5430806348152437|0.0|0.0|1.5430806348152437|
|Cosh|1|64|1.543080634815244|1.4389695516567557e-16|1.4389695516567557e-16|1.5430806348152437|
|---|---|---|---|---|---|---|
|Cosh|2|4|3.6666666666666665|-0.02539182760837448|0.02539182760837448|3.7621956910836314
|Cosh|2|16|3.762195691042252|-1.0998721967992732e-11|1.0998721967992732e-11|3.7621956910836314
|Cosh|2|64|3.762195691083632|1.18039901779312e-16|1.18039901779312e-16|3.7621956910836314





By analyzing the table we can see that what we found here is aligned with what we discussed in class. Using the taylor series polynomial and more n iterations, the relative and absolute errors get smaller and smaller every time. This proves to be correct.